badge_server/main.py

```diff
@@ -137,24 +138,54 @@ def get_highest_status(cls, statuses: Iterable[enum.Enum]):
 }
 
 
-def _get_supported_versions(package_name: str) -> List[int]:
-    """Gets the given package's supported python versions
+def _get_missing_details(package_names: List[str],
+                         results: Iterable[CompatibilityResult]
+                         ) -> Optional[str]:
+    """Gets the details for any missing data (if there is any)
 
     Args:
-        package_name: The name of the package to look up, e.g. "tensorflow".
+        package_names: A list of length 1 or 2 of the package name(s) to look
+            up, e.g. ["tensorflow"], ["tensorflow", "opencensus"].
+        results: A list of length 1 or 2 of the `CompatibilityResults` for the
+            given package(s). Results should come from datastore.
 
     Returns:
-        A list of supported python versions.
+        None if there is no missing data; a description of the version(s)
+        missing otherwise, e.g. "Missing data for python version(s): 2".
     """
-    supported = []
-    if not compat_utils._is_package_in_whitelist([package_name]):
-        return supported
-    unsupported_package_mapping = configs.PKG_PY_VERSION_NOT_SUPPORTED
-    for version in (2, 3):
-        unsupported_packages = unsupported_package_mapping[version]
-        if package_name not in unsupported_packages:
-            supported.append(version)
-    return supported
+    expected_lengths = (1, 2)
+    bad_list_length = 'package_names: Expected length of 1 or 2, got {}'
+    not_whitelisted = 'One of the packages in {} is not whitelisted'
+
+    assert len(package_names) in expected_lengths, bad_list_length.format(
+        len(package_names))
+
+    assert compat_utils._is_package_in_whitelist(
+        package_names), not_whitelisted.format(str(package_names))
+
+    all_versions = (2, 3)
+
+    versions_supported = set(version for version in all_versions)
+    for version in all_versions:
+        for package_name in package_names:
+            if package_name in configs.PKG_PY_VERSION_NOT_SUPPORTED[version]:
+                versions_supported.discard(version)
+
+    versions_seen = {result.python_major_version for result in results}
+
+    if versions_seen.issuperset(versions_supported):
+        return None
+
+    missing_versions = [v for v in versions_supported - versions_seen]
```

Comment:
list(versions_support - versions_seen)
```suggestion
    missing_versions = list(versions_supported - versions_seen)
```

But I don't know why the conversion to list is necessary.

Reply:
The error message without it bothered me - I wanted the presentation to be consistent:
"Missing data for packages=['opencensus'], versions={2, 3}"